src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1312,6 +1312,33 @@ private bool ImportIntrinsicCall(MethodDesc method)
                         return true;
                     }
                     break;
+                case ".ctor":
+                    if (metadataType.IsByReferenceOfT)
+                    {
+                        StackEntry byRefValueParamHolder = _stack.Pop();
+
+                        // Allocate a slot on the shadow stack for the ByReference type
+                        int spillIndex = _spilledExpressions.Count;
+                        SpilledExpressionEntry spillEntry = new SpilledExpressionEntry(StackValueKind.ByRef, "byref" + _currentOffset, metadataType, spillIndex, this);
+                        _spilledExpressions.Add(spillEntry);
+                        LLVMValueRef addrOfValueType = LoadVarAddress(spillIndex, LocalVarKind.Temp, out TypeDesc unused);
+                        var typedAddress = CastIfNecessary(_builder, addrOfValueType, LLVM.PointerType(LLVM.Int32Type(), 0));
+                        LLVM.BuildStore(_builder, byRefValueParamHolder.ValueForStackKind(byRefValueParamHolder.Kind, _builder, false), typedAddress);
+                        AddressExpressionEntry valueTypeByRef = new AddressExpressionEntry(StackValueKind.ByRef, "byref_ctor" + _currentOffset, byRefValueParamHolder.ValueForStackKind(StackValueKind.ByRef, _builder, false));
+
+                        _stack.Push(valueTypeByRef);
+                        return true;
+                    }
+                    break;
+                case "As":
+                    if (metadataType.Namespace == "Internal.Runtime.CompilerServices" && metadataType.Name == "Unsafe")
+                    {
+                        StackEntry valueEntry = _stack.Pop();
+                        LLVMValueRef typedLoadLocation = LoadVarAddress(0, LocalVarKind.Argument, out TypeDesc type);
```

Comment:
This should be loading based on valueEntry, not arg0 of the caller. Also, did you intend to include this in this change?

Reply:
Codegen shouldn't have to special case this - this is regular IL. System.Runtime.CompilerServices.Unsafe is going to have the same IL in this method, for example, but the method is not marked `Intrinsic` there. We only mark this one as Intrinsic because the internal version of Unsafe is used within System.Private.CoreLib and CoreLib is written in C#, the method body cannot be expressed in C#, and CoreLib cannot reference any other assembly. The special handling of internal Unsafe [ends here](https://github.com/dotnet/corert/blob/bdc6bfbef12b94db9dedef90cb39da88949f61ed/src/Common/src/TypeSystem/IL/ILProvider.cs#L57-L62) (where we generate IL for it).

Reply:
Need to investigate this more as without this the test which exercises the ByReference ctor is not passing.  I've made @morganbr's change for now, and will compare the resulting llvm with that which is generated by the internal IL.  Either this change is working round a bug in my ByReference ctor, or there's something else going on.

Reply:
You shouldn't need this line at all -- it's emitting code to load the first argument of the calling function (which isn't necessarily the argument to Unsafe.As<T>). While this discards that load, the type coming out of it is from the caller. The type that you need to push here is the generic argument to Unsafe.As. For an example of how to get ```T```, take a look at the ```get_Value``` intrinsic above.

Reply:
I was following @MichalStrehovsky advice that I shouldn't need this at all, but I'm back to my earlier problem where I need the changes to the `get_Value`.  In short the ctor only works, i.e. passes string concatenation tests, if I have either the `get_Value` change or the Unsafe.As (which is broken as you point out anyway).  I dont know, but maybe both those changes just work around a bug in the ctor.